torch/distributions/utils.py

```diff
@@ -130,19 +132,35 @@ def probs_to_logits(probs, is_binary=False):
     return torch.log(ps_clamped)
 
 
-class lazy_property:
+T = TypeVar("T", contravariant=True)
+R = TypeVar("R", covariant=True)
+
+
+class lazy_property(Generic[T, R]):
```

Comment:
Huh, so we did decide to parameterize based on the input of the callable and the output now?

Reply:
I did this as an alternative to the `ParamSpec` you suggested. I do not think using `ParamSpec` here is wise, because it would allow to decorate arbitrary functions with `@lazy_property`, which doesn't make sense if for instance the function requires extra arguments. Even with optional arguments, I have never seen that and it seems like an ill-advised use-case, so to me it makes most sense to limit the allowed Callables to a single argument ("self")

Reply:
Ah, I misunderstood, the callable is limited to one argument, then this makes sense.